README.md
# Arduino Client for MQTT PlugPlay

This library provides a client for doing simple publish/subscribe messaging with MQTT PlugPlay.

This library is derived almost entirely from the great work done here: https://github.com/knolleary/pubsubclient

## Examples

The library comes with a number of example sketches. See File > Examples > PlugPlayMQTT
within the Arduino application.

## Limitations

 - It can only publish QoS 0 messages. It can subscribe at QoS 0 or QoS 1.
 - The maximum message size, including header, is **128 bytes** by default. This
   is configurable via `MQTT_MAX_PACKET_SIZE` in `PubSubClient.h`.
 - The keepalive interval is set to 15 seconds by default. This is configurable
   via `MQTT_KEEPALIVE` in `PubSubClient.h`.
 - The client uses MQTT 3.1.1 by default. It can be changed to use MQTT 3.1 by
   changing value of `MQTT_VERSION` in `PubSubClient.h`.

## Compatible Hardware

The library uses the Arduino Ethernet Client api for interacting with the
underlying network hardware. This means it Just Works with a growing number of
boards and shields, including:

 - Arduino Ethernet
 - Arduino Ethernet Shield
 - Arduino YUN – use the included `YunClient` in place of `EthernetClient`, and
   be sure to do a `Bridge.begin()` first
 - Arduino WiFi Shield - if you want to send packets > 90 bytes with this shield,
   enable the `MQTT_MAX_TRANSFER_SIZE` define in `PubSubClient.h`.
 - Sparkfun WiFly Shield – [library](https://github.com/dpslwk/WiFly)
 - TI CC3000 WiFi - [library](https://github.com/sparkfun/SFE_CC3000_Library)
 - Intel Galileo/Edison
 - ESP8266

The library cannot currently be used with hardware based on the ENC28J60 chip –
such as the Nanode or the Nuelectronics Ethernet Shield. For those, there is an
[alternative library](https://github.com/njh/NanodeMQTT) available.

## Install

Clone (or download and unzip) the repository to `~/Documents/Arduino/libraries`
where `~/Documents/Arduino` is your sketchbook directory.

    > cd ~/Documents/Arduino
    > mkdir libraries
    > cd libraries
    > git clone https://github.com/plugplayco/plugplay-arduino-mqtt PlugPlayMQTT

## Usage

### setAuth( const char* userKey, const char* boardId)

Set authentication to connect to PlugPlay

Arguments: 
- *userKey* - user Key of your user account.
- *boardId* - board Id of board you want to publish/subcribe to.

### boolean connectPlugPlay(const char* id)

Connect to PlugPlay with a MQTT client Id

Arguments: 
- *id* - client Id (should be a random Id)

```c++
// Create a random client ID
String clientId = "ArduinoClient-";
clientId += String(random(0xffff), HEX);

client.connectPlugPlay(clientId.c_str());
```

### char* createMsg(String dvName, float data0)
### char* createMsg(String dvName, float data0, float data1)
### char* createMsg(String dvName, float data0, float data1, float data2)

Create a message to publish

Arguments: 
- *dvName* - device name
- *data0* - first data
- *data1* - second data
- *data2* - third data

### boolean subPlugPlay(const char* topic)

Subcribe a topic 

### boolean pubPlugPlay(const char* topic, const char* payload)

Publish a message to a topic. The message is created by *createMsg(String dvName, float data0, float data1, float data2)*

### String getName(const char* payload)

Get device name from a received message

### float getData(const char* payload, const char* type)

Get data from a received message

```c++
// Get the first data
client.getData((char*)payload,"data0");
// Get the second data
client.getData((char*)payload,"data1");
// Get the third data
client.getData((char*)payload,"data2");
```

### Other functions

You can refer here: http://pubsubclient.knolleary.net 

## License

This code is released under the MIT License.


CHANGES.txt
/************************ PlugPlayMQTT ***********************/

1.0
    * Addd ArduinoJson Lib at commit bff77ab 
    * Add PlugPlayMQTT class
    * Edit examples code
    * Edit info files

/************************ PubSubClient ***********************/

2.4
   * Add MQTT_SOCKET_TIMEOUT to prevent it blocking indefinitely
     whilst waiting for inbound data
   * Fixed return code when publishing >256 bytes
   
2.3
   * Add publish(topic,payload,retained) function

2.2
   * Change code layout to match Arduino Library reqs

2.1
   * Add MAX_TRANSFER_SIZE def to chunk messages if needed
   * Reject topic/payloads that exceed MQTT_MAX_PACKET_SIZE

2.0
   * Add (and default to) MQTT 3.1.1 support
   * Fix PROGMEM handling for Intel Galileo/ESP8266
   * Add overloaded constructors for convenience
   * Add chainable setters for server/callback/client/stream
   * Add state function to return connack return code

1.9
   * Do not split MQTT packets over multiple calls to _client->write()
   * API change: All constructors now require an instance of Client
      to be passed in.
   * Fixed example to match 1.8 api changes - dpslwk
   * Added username/password support - WilHall
   * Added publish_P - publishes messages from PROGMEM - jobytaffey

1.8
    * KeepAlive interval is configurable in PubSubClient.h
    * Maximum packet size is configurable in PubSubClient.h
    * API change: Return boolean rather than int from various functions
    * API change: Length parameter in message callback changed
       from int to unsigned int
    * Various internal tidy-ups around types
1.7
    * Improved keepalive handling
    * Updated to the Arduino-1.0 API
1.6
    * Added the ability to publish a retained message

1.5
    * Added default constructor
    * Fixed compile error when used with arduino-0021 or later

1.4
    * Fixed connection lost handling

1.3
    * Fixed packet reading bug in PubSubClient.readPacket

1.2
    * Fixed compile error when used with arduino-0016 or later


1.1
    * Reduced size of library
    * Added support for Will messages
    * Clarified licensing - see LICENSE.txt


1.0
    * Only Quality of Service (QOS) 0 messaging is supported
    * The maximum message size, including header, is 128 bytes
    * The keepalive interval is set to 30 seconds
    * No support for Will messages


src/PlugPlayMQTT.h
#ifndef PlugPlayMQTT_h
#define PlugPlayMQTT_h

#include <PubSubClient.h>
#include <ArduinoJson.h>

// MQTT_MAX_PAYLOAD_SIZE : Maximum payload size inside a packet
#ifndef MQTT_MAX_PAYLOAD_SIZE
#define MQTT_MAX_PAYLOAD_SIZE 100
#endif

class PlugPlayMQTT: public PubSubClient {

    private:
        const char* _userKey; 
        const char* _boardId;
        String _authSubTopic, _authPubTopic;
        char payload[MQTT_MAX_PAYLOAD_SIZE];
    public:
        //PlugPlayMQTT(Client& client):PubSubClient(client){};
        using PubSubClient::PubSubClient;
        PlugPlayMQTT& setAuth( const char* userKey, const char* boardId);
        boolean connectPlugPlay(const char* id);
        boolean subPlugPlay(const char* topic);
        boolean pubPlugPlay(const char* topic, const char* payload);
        
        char* createMsg(String dvName, float data0);
        char* createMsg(String dvName, float data0, float data1);
        char* createMsg(String dvName, float data0, float data1, float data2);
        
        String   getName(const char* payload);
        float    getData(const char* payload, const char* type);
};

#endif

src/PlugPlayMQTT.cpp
#include "PlugPlayMQTT.h"
#include "Arduino.h"

PlugPlayMQTT& PlugPlayMQTT::setAuth( const char* userKey, const char* boardId) {
    this->_userKey = userKey;
    this->_boardId = boardId;
    return *this;
}

boolean PlugPlayMQTT::connectPlugPlay(const char* id) {
    //Note: In connect(), when passw is available, user has to be available (!= NULL)
    return connect(id,_userKey,_boardId);
}

boolean PlugPlayMQTT::subPlugPlay(const char* topic) {
    _authSubTopic = String(_boardId) + "/" + String(topic);
    return subscribe(_authSubTopic.c_str(), 0);
}

boolean PlugPlayMQTT::pubPlugPlay(const char* topic, const char* payload) {
    _authPubTopic = String(_boardId) + "/" + String(topic);
    return publish(_authPubTopic.c_str(),(const uint8_t*)payload,strlen(payload),false);
}

char* PlugPlayMQTT::createMsg(String dvName, float data0){    
    return createMsg(dvName, data0, 0, 0);
}

char* PlugPlayMQTT::createMsg(String dvName, float data0, float data1){    
    return createMsg(dvName, data0, data1, 0);
}

char* PlugPlayMQTT::createMsg(String dvName, float data0, float data1, float data2){    
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.createObject(); 
    root["device"] = dvName;
    root["data0"] = data0;
    root["data1"] = data1;
    root["data2"] = data2;
    root.printTo(payload, MQTT_MAX_PAYLOAD_SIZE);
    return payload;
}

String PlugPlayMQTT::getName(const char* payload){
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.parseObject(payload);
    if (!root.success()) {
        Serial.println("parseMsg() failed");
        return "";
    }
    return root["device"];
} 

float PlugPlayMQTT::getData(const char* payload, const char* type){
    StaticJsonBuffer<MQTT_MAX_PAYLOAD_SIZE> jsonBuffer;
    JsonObject& root = jsonBuffer.parseObject(payload);
    if (!root.success()) {
        Serial.println("parseMsg() failed");
        return 0;
    }    
    return root[type];
}

src/PubSubClient.h
/*
 PubSubClient.h - A simple client for MQTT.
  Nick O'Leary
  http://knolleary.net
*/

#ifndef PubSubClient_h
#define PubSubClient_h

#include <Arduino.h>
#include "IPAddress.h"
#include "Client.h"
#include "Stream.h"

#define MQTT_VERSION_3_1      3
#define MQTT_VERSION_3_1_1    4

// MQTT_VERSION : Pick the version
//#define MQTT_VERSION MQTT_VERSION_3_1
#ifndef MQTT_VERSION
#define MQTT_VERSION MQTT_VERSION_3_1_1
#endif

// MQTT_MAX_PACKET_SIZE : Maximum packet size
#ifndef MQTT_MAX_PACKET_SIZE
#define MQTT_MAX_PACKET_SIZE 128
#endif

// MQTT_KEEPALIVE : keepAlive interval in Seconds
#ifndef MQTT_KEEPALIVE
#define MQTT_KEEPALIVE 15
#endif

// MQTT_SOCKET_TIMEOUT: socket timeout interval in Seconds
#ifndef MQTT_SOCKET_TIMEOUT
#define MQTT_SOCKET_TIMEOUT 15
#endif

// MQTT_MAX_TRANSFER_SIZE : limit how much data is passed to the network client
//  in each write call. Needed for the Arduino Wifi Shield. Leave undefined to
//  pass the entire MQTT packet in each write call.
//#define MQTT_MAX_TRANSFER_SIZE 80

// Possible values for client.state()
#define MQTT_CONNECTION_TIMEOUT     -4
#define MQTT_CONNECTION_LOST        -3
#define MQTT_CONNECT_FAILED         -2
#define MQTT_DISCONNECTED           -1
#define MQTT_CONNECTED               0
#define MQTT_CONNECT_BAD_PROTOCOL    1
#define MQTT_CONNECT_BAD_CLIENT_ID   2
#define MQTT_CONNECT_UNAVAILABLE     3
#define MQTT_CONNECT_BAD_CREDENTIALS 4
#define MQTT_CONNECT_UNAUTHORIZED    5

#define MQTTCONNECT     1 << 4  // Client request to connect to Server
#define MQTTCONNACK     2 << 4  // Connect Acknowledgment
#define MQTTPUBLISH     3 << 4  // Publish message
#define MQTTPUBACK      4 << 4  // Publish Acknowledgment
#define MQTTPUBREC      5 << 4  // Publish Received (assured delivery part 1)
#define MQTTPUBREL      6 << 4  // Publish Release (assured delivery part 2)
#define MQTTPUBCOMP     7 << 4  // Publish Complete (assured delivery part 3)
#define MQTTSUBSCRIBE   8 << 4  // Client Subscribe request
#define MQTTSUBACK      9 << 4  // Subscribe Acknowledgment
#define MQTTUNSUBSCRIBE 10 << 4 // Client Unsubscribe request
#define MQTTUNSUBACK    11 << 4 // Unsubscribe Acknowledgment
#define MQTTPINGREQ     12 << 4 // PING Request
#define MQTTPINGRESP    13 << 4 // PING Response
#define MQTTDISCONNECT  14 << 4 // Client is Disconnecting
#define MQTTReserved    15 << 4 // Reserved

#define MQTTQOS0        (0 << 1)
#define MQTTQOS1        (1 << 1)
#define MQTTQOS2        (2 << 1)

#ifdef ESP8266
#include <functional>
#define MQTT_CALLBACK_SIGNATURE std::function<void(char*, uint8_t*, unsigned int)> callback
#else
#define MQTT_CALLBACK_SIGNATURE void (*callback)(char*, uint8_t*, unsigned int)
#endif

class PubSubClient {
protected:
   Client* _client;
   uint8_t buffer[MQTT_MAX_PACKET_SIZE];
   uint16_t nextMsgId;
   unsigned long lastOutActivity;
   unsigned long lastInActivity;
   bool pingOutstanding;
   MQTT_CALLBACK_SIGNATURE;
   uint16_t readPacket(uint8_t*);
   boolean readByte(uint8_t * result);
   boolean readByte(uint8_t * result, uint16_t * index);
   boolean write(uint8_t header, uint8_t* buf, uint16_t length);
   uint16_t writeString(const char* string, uint8_t* buf, uint16_t pos);
   IPAddress ip;
   const char* domain;
   uint16_t port;
   Stream* stream;
   int _state;
public:
   PubSubClient();
   PubSubClient(Client& client);
   PubSubClient(IPAddress, uint16_t, Client& client);
   PubSubClient(IPAddress, uint16_t, Client& client, Stream&);
   PubSubClient(IPAddress, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client);
   PubSubClient(IPAddress, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client, Stream&);
   PubSubClient(uint8_t *, uint16_t, Client& client);
   PubSubClient(uint8_t *, uint16_t, Client& client, Stream&);
   PubSubClient(uint8_t *, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client);
   PubSubClient(uint8_t *, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client, Stream&);
   PubSubClient(const char*, uint16_t, Client& client);
   PubSubClient(const char*, uint16_t, Client& client, Stream&);
   PubSubClient(const char*, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client);
   PubSubClient(const char*, uint16_t, MQTT_CALLBACK_SIGNATURE,Client& client, Stream&);

   PubSubClient& setServer(IPAddress ip, uint16_t port);
   PubSubClient& setServer(uint8_t * ip, uint16_t port);
   PubSubClient& setServer(const char * domain, uint16_t port);
   PubSubClient& setCallback(MQTT_CALLBACK_SIGNATURE);
   PubSubClient& setClient(Client& client);
   PubSubClient& setStream(Stream& stream);

   boolean connect(const char* id);
   boolean connect(const char* id, const char* user, const char* pass);
   boolean connect(const char* id, const char* willTopic, uint8_t willQos, boolean willRetain, const char* willMessage);
   boolean connect(const char* id, const char* user, const char* pass, const char* willTopic, uint8_t willQos, boolean willRetain, const char* willMessage);
   void disconnect();
   boolean publish(const char* topic, const char* payload);
   boolean publish(const char* topic, const char* payload, boolean retained);
   boolean publish(const char* topic, const uint8_t * payload, unsigned int plength);
   boolean publish(const char* topic, const uint8_t * payload, unsigned int plength, boolean retained);
   boolean publish_P(const char* topic, const uint8_t * payload, unsigned int plength, boolean retained);
   boolean subscribe(const char* topic);
   boolean subscribe(const char* topic, uint8_t qos);
   boolean unsubscribe(const char* topic);
   boolean loop();
   boolean connected();
   int state();
};


#endif
